contrib/node/src/python/pants/contrib/node/subsystems/eslint_distribution.py

```diff
@@ -0,0 +1,133 @@
+      return ESLintDistribution(options.setupdir, options.supportdir, options.eslint_config,
+                                options.eslint_ignore, pants_bootstrapdir)
+
+  @property
+  def setupdir(self):
+    return self._setupdir
+
+  @property
+  def supportdir(self):
+    return self._supportdir
+
+  @property
+  def eslint_config(self):
+    return self._eslint_config
+
+  @property
+  def eslint_ignore(self):
+    return self._eslint_ignore
+
+  @property
+  def pants_bootstrapdir(self):
+    return self._pants_bootstrapdir
+
+  def __init__(self, setupdir, supportdir, eslint_config, eslint_ignore, pants_bootstrapdir):
+    self._setupdir = setupdir
+    self._supportdir = supportdir
+    self._eslint_config = eslint_config or None
+    self._eslint_ignore = eslint_ignore or None
+    self._pants_bootstrapdir = pants_bootstrapdir
+
+  def _is_workingdir_valid(self, workingdir):
+    dir_exists = os.path.isdir(workingdir)
+    if not dir_exists:
+      return False
+    else:
+      lock_file = os.path.join(workingdir, 'yarn.lock')
+      package_json = os.path.join(workingdir, 'package.json')
+      if not os.path.isfile(package_json):
+        logger.warning('The workingdir: `{}` is missing a package.json'.format(workingdir))
+        return False
+      elif not os.path.isfile(lock_file):
+        logger.warning('A lock file was not found, consider adding a yarn.lock file for '
+                       'deterministic installation.')
+    return True
+
+  def _compare_file_checksums(self, file_a, file_b):
+    if os.path.isfile(file_a) and os.path.isfile(file_b):
+      return self._checksum_md5(file_a) == self._checksum_md5(file_b)
+    return False
+
+  def _checksum_md5(self, filename):
+    md5 = hashlib.md5()
+    with open(filename, 'rb') as f:
+      for block in iter(lambda: f.read(128 * md5.block_size), b''):
+        md5.update(block)
+    return md5.hexdigest()
+
+  def _bootstrap_eslinter(self, bootstrapped_support_path):
+    if self._setupdir is None or not self._is_workingdir_valid(self._setupdir):
+      safe_mkdir(bootstrapped_support_path)
+      return False
+    logger.debug('Copying {setupdir} to bootstrapped dir: {support_path}'
+                           .format(setupdir=self._setupdir, support_path=bootstrapped_support_path))
+    safe_rmtree(bootstrapped_support_path)
+    shutil.copytree(self._setupdir, bootstrapped_support_path)
+    return True
+
+  def fetch_supportdir(self):
+    """ Returns the path where the ESLintDistribution is bootstrapped.
+
+    :returns: The path where ESLintDistribution is bootstrapped and whether or not it was preconfigured
+    :rtype: (string, bool)
+    """
+    bootstrap_dir = os.path.realpath(os.path.expanduser(self._pants_bootstrapdir))
+    bootstrapped_support_path = os.path.join(bootstrap_dir, self._supportdir)
+    is_preconfigured = True
+    if not os.path.exists(bootstrapped_support_path):
+     is_preconfigured = self._bootstrap_eslinter(bootstrapped_support_path)
```

Comment:
should we consider to remove preconfigured bootstrap dir during `pants clean-all`? otherwise, if we downloaded a bad file/configure, users can't easily delete them and retry.
I think a TODO for now is OK, but we should try to fix it soon.

Reply:
Agreed, added a TODO here as the clean-up should clean up other things as well.